Striped improvements

mike-barber opened the pull request:
- corrected extra iteration on main sieve
- added more cache-friendly `FlagStorageBitVectorStripedBlocks` version of the striped implementation
- simplified stripe/word accounting on `FlagStorageBitVectorStriped` based on learnings from the above
- basic tests added to validate storage directly

@Kulasko this might be of interest to you for your parallel implementations -- you could refactor this a bit and run the different blocks in parallel on different threads. 

It's a lot faster on the raspberri pi 4 as well -- I think it might be a bit quicker on your machine too.

mike-barber committed:
added striped block storage + test

mike-barber committed:
added options

mike-barber committed:
added unrolled

mike-barber committed:
unrolled 4x

mike-barber committed:
early termination fixes

mike-barber committed:
cleanup

mike-barber committed:
const generic version

mike-barber committed:
cleanup

mike-barber committed:
cleanup and comments

mike-barber committed:
cleanup

Kulasko commented:
Holy cow, now that is fast! As you know, bit_rotate has a performance ceiling of about 4500 in my machine, while the original striped version results in about 3300 passes. The big block achieve about 4400 passes, while the striped one with small blocks got almost 8000 passes! This easily makes it the fastest base algorithm one on my machine, by far.

Kulasko commented:
A note on idiomatic code, though:
While the inner loop needs to set the initial bit index, the outer loop in `reset_flags` for `FlagStorageBitVectorStripedBlocks` has no such dependency and therefore can use an iterator: `for block_idx in start / Self::BLOCK_SIZE_BITS..self.blocks.len() { ... }`
This should compile down to the exact same code since the iterator is simple. On my machine, it makes no difference or even has a small positive impact. On the other hand, most of the other code uses while loops now, so one might prefer the consistency.
Also since Rust 1.51 (I believe), intra-doc links are supported. That means that for example, your can directly link to `FlagStorageBitVectorStriped` from the documentation of `FlagStorageBitVectorStripedBlocks` (the name begins to remind me of C# :P) by simply declaring it as a link: ```[`FlagStorageBitVectorStriped`]```.
I also would like to see a short sentence for both `BLOCK_SIZE_DEFAULT` and `BLOCK_SIZE_SMALL` if that is alright with you.

mike-barber committed:
idiomatic quality improvements - thanks @Kulasko

mike-barber committed:
updated perf numbers in docs

mike-barber commented:
Wow that's really fast on your machine. The cache locality really makes a difference :) The raspberri pi really likes it too, especially the smaller block size with the multithreaded workload -- it's able so scale a lot better than previously with more threads. 

```
mike-barber_byte-storage;2614;5.0016665459;1;algorithm=base,faithful=yes,bits=8
mike-barber_bit-storage;2179;5.0017485619;1;algorithm=base,faithful=yes,bits=1
mike-barber_bit-storage-rotate;2079;5.0009078979;1;algorithm=base,faithful=yes,bits=1
mike-barber_bit-storage-striped;2422;5.0011348724;1;algorithm=base,faithful=yes,bits=1
mike-barber_bit-storage-striped-blocks;3421;5.0014257431;1;algorithm=base,faithful=yes,bits=1
mike-barber_bit-storage-striped-blocks-small;3434;5.0014367104;1;algorithm=base,faithful=yes,bits=1
mike-barber_byte-storage;398;5.0321965218;4;algorithm=base,faithful=yes,bits=8
mike-barber_bit-storage;7501;5.0025444031;4;algorithm=base,faithful=yes,bits=1
mike-barber_bit-storage-rotate;7075;5.0023360252;4;algorithm=base,faithful=yes,bits=1
mike-barber_bit-storage-striped;4267;5.0038175583;4;algorithm=base,faithful=yes,bits=1
mike-barber_bit-storage-striped-blocks;9626;5.0019483566;4;algorithm=base,faithful=yes,bits=1
mike-barber_bit-storage-striped-blocks-small;11277;5.0010590553;4;algorithm=base,faithful=yes,bits=1
```

> While the inner loop needs to set the initial bit index, the outer loop in reset_flags for FlagStorageBitVectorStripedBlocks has no such dependency and therefore can use an iterator:

Thanks -- that's a really good call. I've updated it. 

> Also since Rust 1.51 (I believe), intra-doc links are supported. That means that for example, your can directly link to FlagStorageBitVectorStriped from the documentation of FlagStorageBitVectorStripedBlocks (the name begins to remind me of C# :P) by simply declaring it as a link: [`FlagStorageBitVectorStriped`].

That is very cool indeed. I didn't know about that. The links even work in VS Code with rust-analyzer :) 

> I also would like to see a short sentence for both BLOCK_SIZE_DEFAULT and BLOCK_SIZE_SMALL if that is alright with you.

Agreed -- I've added a little note to each to explain what they're for.

Thanks for this. Let me know if it looks good :)

Kulasko commented:
Looks good so far. :)
Just out of curiosity, did you test the algorithm with 8192 bytes and/or u32 elements?

Kulasko commented:
Just tested this, seems you indeed found the best combination of things.

mike-barber commented:
I did :)

8k byte blocks were a little slower than 16k on my machine, and the same applies to larger blocks (32k and up). With 64k, the performance is the about same as the original striped case, because the whole thing fits in a single block. 

Although it's usually faster dealing with u32 elements, it is quite slow in the striped algorithm because we land up doing 32 loops instead of 8, and the overheads start adding up.

> Just tested this, seems you indeed found the best combination of things.

Thanks! I'm glad it's not just tuned to just my machine. Hopefully they're testing on a Ryzen/Threadripper machine though :)

I'll open a PR for `drag-race` :)
